storagenode/pieces/store_test.go

```diff
@@ -181,6 +181,57 @@ func tryOpeningAPiece(ctx context.Context, t testing.TB, store *pieces.Store, sa
 	require.NoError(t, reader.Close())
 }
 
+func TestPieceVersionMigrate(t *testing.T) {
+	storagenodedbtest.Run(t, func(t *testing.T, db storagenode.DB) {
+		const pieceSize = 1024
+
+		ctx := testcontext.New(t)
+		defer ctx.Cleanup()
+
+		var (
+			data      = testrand.Bytes(pieceSize)
+			satellite = testrand.NodeID()
+			pieceID   = testrand.PieceID()
+			now       = time.Now().UTC()
+		)
+
+		blobs, err := filestore.NewAt(zaptest.NewLogger(t), ctx.Dir("store"))
+		require.NoError(t, err)
+		defer ctx.Check(blobs.Close)
+
+		pieceInfo := pieces.Info{
+			SatelliteID:     satellite,
+			PieceID:         pieceID,
+			OrderLimit:      &pb.OrderLimit{},
+			PieceCreation:   now,
+			UplinkPieceHash: &pb.PieceHash{},
```

Comment:
Would it be too much of a pain to have a 'real' signed `PieceHash` and `OrderLimit` inserted here, so that we can verify later that the signatures and hashes still correctly validate after the v0->v1 transformation? It's pretty important that those come through intact. I don't know why they wouldn't, but we should really check.

It looks like something similar is done fairly simply in pkg/signing/sign_test.go, if that helps.

Reply:
I added a dummy signature for the satellite and piece hash, and then verified that the values are the same in the v0 and v1 pieces, with the assumption that if the values are the same the signatures would be valid in both cases. (I'm not using real signatures, just dummy ones.)

I also am verifying that `PieceCreation` value is correct.

Let me know what you think of this approach.

Reply:
That ought to be good enough. Cool.

Reply:
First, we don't store an entire `PieceHash` object in v1 pieces, we only take the signature value from the piece hash and place it in the `PieceHeader`. So for that one, only the signature value matters.

For the OrderLimit, I added many more fields and confirmed that the entire OrderLimit object remains unchanged.

Reply:
Oh no that's what I mean by taking apart the PieceHash and putting it back together. We actually _do_ store the other parts of a PieceHash, but in different ways. Piece ID and piece size come from the file system. Timestamp and Hash are stored as CreationTime and Hash in the piece header. It's imperative that we be able to reconstruct a PieceHash exactly in order to validate the signature.

Reply:
I've updated the test to sign the v0 piece hash, and validate the signature of the v1 piece hash. I'm manually creating the piecehash from the v1 info (similar to https://github.com/storj/storj/blob/master/storagenode/piecestore/endpoint.go#L514).

The tests do pass, it is indeed able to validate the signature with the v1 information after the migration.